# Exercise 2
**Name**: Francis O'Hara Aidoo
**Roll Number**: 10201100101 <br>

Task: Solve all practice questions from the following webpages:

1.https://pythonbasics.org/pandas-dataframe/

1.1 https://data36.com/pandas-tutorial-1-basics-reading-data-files-dataframes-data-selection/

2.https://www.tutorialspoint.com/python_pandas/python_pandas_dataframe.htm
3.https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/   (different sources)

# Website 1 Solutions

#### Creating n x 1 Dataframe from List
The following code creates an n x 1 dataframe from a list containing n elements.

In [1]:
import pandas as pd

fruits = ["Banana", "Apple", "Guava", "Straw berry"]
fruits_df = pd.DataFrame(fruits)

print(fruits_df)

             0
0       Banana
1        Apple
2        Guava
3  Straw berry


### Creating m x n Dataframe from List
The following code creates an m x n dataframe from a list of m lists each of length n

In [2]:
import pandas as pd

grades = [["Anna Tetteh", 88, "B2"], ["Eugene Stark", 54, "D7"], ["Helen Troy", 100, "A1+"], ["Bruce Banner", 92, "A1"]]
column_labels = ["Name", "Score", "Grade"]
grades_df = pd.DataFrame(grades, columns=column_labels)
print(grades_df)

           Name  Score Grade
0   Anna Tetteh     88    B2
1  Eugene Stark     54    D7
2    Helen Troy    100   A1+
3  Bruce Banner     92    A1


### Selecting column from a DataFrame by column name
The following code selects the Name column from the grades_df DataFrame we created in the previous cell.

In [3]:
names = grades_df["Name"]
print(names)

0     Anna Tetteh
1    Eugene Stark
2      Helen Troy
3    Bruce Banner
Name: Name, dtype: object


### Adding columns to an existing DataFrame
The following code adds a column from a newly created DataFrame to an existing DataFrame.

In [4]:
# Grades_df before addition of DataFrame
print("----------------Before-Addition------------------")
print(grades_df)

courses_df = pd.DataFrame(["Information Technology", "Computer Science", "Computer Engineering", "Business"], columns=["Course"]) # New DataFrame
grades_df["Course"] = courses_df["Course"]

# Grades_df after addition of DataFrame
print("-----------------After-Addition-------------------")
print(grades_df)

----------------Before-Addition------------------
           Name  Score Grade
0   Anna Tetteh     88    B2
1  Eugene Stark     54    D7
2    Helen Troy    100   A1+
3  Bruce Banner     92    A1
-----------------After-Addition-------------------
           Name  Score Grade                  Course
0   Anna Tetteh     88    B2  Information Technology
1  Eugene Stark     54    D7        Computer Science
2    Helen Troy    100   A1+    Computer Engineering
3  Bruce Banner     92    A1                Business


### Deleting columns
The following code snippet deletes the Course column from grades_df using the <em>del</em> Python keyword

In [5]:
print("-----------------Before-Deletion-------------------")
print(grades_df)
del grades_df["Course"]
print("-----------------After-Deletion-------------------")
print(grades_df)

-----------------Before-Deletion-------------------
           Name  Score Grade                  Course
0   Anna Tetteh     88    B2  Information Technology
1  Eugene Stark     54    D7        Computer Science
2    Helen Troy    100   A1+    Computer Engineering
3  Bruce Banner     92    A1                Business
-----------------After-Deletion-------------------
           Name  Score Grade
0   Anna Tetteh     88    B2
1  Eugene Stark     54    D7
2    Helen Troy    100   A1+
3  Bruce Banner     92    A1


### Selecting single rows from data frame using .loc method
The following code snippet selects the third row from the grades_df data frame.

In [6]:
row3 = grades_df.loc[2]
print(row3)

Name     Helen Troy
Score           100
Grade           A1+
Name: 2, dtype: object


### Selecting single rows from data frame using .iloc method
The following code snippet selects the 2nd row from the grades_df data frame using the .iloc data frame method.

In [7]:
row2 = grades_df.iloc[1]
print(row2)

Name     Eugene Stark
Score              54
Grade              D7
Name: 1, dtype: object


### Appending rows to a data frame
The following code snippet appends rows from a new ddta frame into the grades_df data frame using pd.concat() method.

In [8]:
new_df = pd.DataFrame([["Bertholdt Hoover", 87, "C3"], ["Armin Arlelt", 100, "A1++"], ["Pieck Finger", 95, "A1"]], columns=["Name", "Score", "Grade"])

print("-----------------Before-Append-------------------")
print(grades_df)
grades_df = pd.concat((grades_df, new_df), ignore_index=True)

print("-----------------After-Append-------------------")
print(grades_df)

-----------------Before-Append-------------------
           Name  Score Grade
0   Anna Tetteh     88    B2
1  Eugene Stark     54    D7
2    Helen Troy    100   A1+
3  Bruce Banner     92    A1
-----------------After-Append-------------------
               Name  Score Grade
0       Anna Tetteh     88    B2
1      Eugene Stark     54    D7
2        Helen Troy    100   A1+
3      Bruce Banner     92    A1
4  Bertholdt Hoover     87    C3
5      Armin Arlelt    100  A1++
6      Pieck Finger     95    A1


### Deleting single rows from a data frame
The following code snippet deletes a single row from the data frame using the .drop() data frame method

In [9]:
print("-----------------Before-deletion-------------------")
print(grades_df)

grades_df = grades_df.drop(1)

print("-----------------After-deletion-------------------")
print(grades_df)

-----------------Before-deletion-------------------
               Name  Score Grade
0       Anna Tetteh     88    B2
1      Eugene Stark     54    D7
2        Helen Troy    100   A1+
3      Bruce Banner     92    A1
4  Bertholdt Hoover     87    C3
5      Armin Arlelt    100  A1++
6      Pieck Finger     95    A1
-----------------After-deletion-------------------
               Name  Score Grade
0       Anna Tetteh     88    B2
2        Helen Troy    100   A1+
3      Bruce Banner     92    A1
4  Bertholdt Hoover     87    C3
5      Armin Arlelt    100  A1++
6      Pieck Finger     95    A1


### Creating a data frame from a dictionary
The following code snippet creates a data frame from a python dictionary.

In [10]:
dictionary = {"ID":["001", "002", "003", "004", "005"], "first_name":["Michael", "Gabriel", "Evangelista", "Hillary", "Sarah"], "last_name":["Asante", "Io", "Neptunes", "Yu", "Pennyworth"], "team":["Alpha", "Bravo", "Charlie", "Delta", "Echo"], "country":["Canada", "USA", "Namibia", "Philippines", "Greenland"]}

soldiers = pd.DataFrame(dictionary)

print(soldiers)

    ID   first_name   last_name     team      country
0  001      Michael      Asante    Alpha       Canada
1  002      Gabriel          Io    Bravo          USA
2  003  Evangelista    Neptunes  Charlie      Namibia
3  004      Hillary          Yu    Delta  Philippines
4  005        Sarah  Pennyworth     Echo    Greenland


### Creating a data frame from a numpy array
The following code snippet defines a numpy array and instantiates a pandas data frame from it:

In [11]:
import numpy as np

numpy_array = np.array([["001", "Michael", "Asante", "Alpha", "Country"], ["002", "Gabriel", "Io", "Bravo", "USA"], ["003", "Evangelista", "Neptunes", "Charlie", "Namibia"], ["004", "Hillary", "Yu", "Delta", "Philippines"], ["005", "Sarah", "Pennyworth", "Echo", "Greenland"]])

new_df2 = pd.DataFrame(numpy_array, columns=["ID", "first_name", "last_name", "team", "country"])

print(new_df2)

    ID   first_name   last_name     team      country
0  001      Michael      Asante    Alpha      Country
1  002      Gabriel          Io    Bravo          USA
2  003  Evangelista    Neptunes  Charlie      Namibia
3  004      Hillary          Yu    Delta  Philippines
4  005        Sarah  Pennyworth     Echo    Greenland


### Creating a data frame from an existing data frame
The following code snippet copies an existing data frame to create a new data frame.

In [12]:
new_df3 = new_df2.copy()
print(new_df3)

    ID   first_name   last_name     team      country
0  001      Michael      Asante    Alpha      Country
1  002      Gabriel          Io    Bravo          USA
2  003  Evangelista    Neptunes  Charlie      Namibia
3  004      Hillary          Yu    Delta  Philippines
4  005        Sarah  Pennyworth     Echo    Greenland


### Creating a data frame from a csv file
The following code snippet creates a data frame by importing data from a csv file:

In [13]:
new_df4 = pd.read_csv("C://users/franc/PycharmProjects/intro-to-data-science/data/data.csv")
print(new_df4)

   id  first_name   last_name      team  country
0   0      Gideon      Asante       Red    Ghana
1   1      Dmitri   Dvocovich      Blue   Russia
2   2    Francois    Arouette     Green   France
3   3     Bradley      Cooper    Purple      USA
4   4      Kratos      Athens   Crimson   Greece


# Website 2 Solutions

### Reading data from a csv file into a data frame in Pandas.
The following code snippet reads data from a .csv file into a pandas data frame.

In [14]:
fruits = pd.read_csv("C://users/franc/PycharmProjects/intro-to-data-science/data/fruits.csv", delimiter=",")
print(fruits)

   id          item   price
0   1         Apple    2.50
1   2        Banana    4.50
2   3   Straw Berry   10.00
3   4     Pineapple    3.75
4   5        Orange    8.00
5   6       Coconut    7.50
6   7         Mango   10.00
7   8         Guava    7.50
8   9         Grape    1.50


# Website 2 Solutions

### Reading Data from a CSV file into a pandas data frame.
The following code snippet reads data from a CSV file specified by the url into a pandas data frame and stores the result in the <em>data</em> variable.

In [15]:
data = pd.read_csv("http://46.101.230.157/dilan/pandas_tutorial_read.csv", delimiter=";")
print(data)

      2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
0     2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
1     2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
2     2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
3     2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
4     2018-01-01 00:05:42  read  country_6  2458151266   Reddit  North America
...                   ...   ...        ...         ...      ...            ...
1789  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1790  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1791  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1792  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1793  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia

[1794 rows x 6 columns]


In the above code snippet, no column labels are included in the csv file. The code snippet below reads the csv file and includes the appropriate column labels as an argument to the pd.read_csv() method.

In [16]:
data = pd.read_csv("http://46.101.230.157/dilan/pandas_tutorial_read.csv", delimiter=";", names=['my_datetime', 'event', 'country', 'user_id', 'source', 'topic'])
print(data)

              my_datetime event    country     user_id   source          topic
0     2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
1     2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
2     2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
3     2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
4     2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
...                   ...   ...        ...         ...      ...            ...
1790  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1792  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1793  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1794  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia

[1795 rows x 6 columns]


### Displaying a sample of your data frame.
The following code snippet uses the head() data frame method to display the first 10 observations in the data set.

In [17]:
print(data.head(10))

           my_datetime event    country     user_id   source          topic
0  2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
1  2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
2  2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
3  2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
4  2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
5  2018-01-01 00:05:42  read  country_6  2458151266   Reddit  North America
6  2018-01-01 00:06:06  read  country_2  2458151267   Reddit         Europe
7  2018-01-01 00:06:15  read  country_6  2458151268  AdWords         Europe
8  2018-01-01 00:07:21  read  country_7  2458151269  AdWords  North America
9  2018-01-01 00:07:29  read  country_5  2458151270   Reddit  North America


The following code snippet uses the tail() data frame method to display the last 10 observations in the data set.

In [18]:
print(data.tail(10))

              my_datetime event    country     user_id   source          topic
1785  2018-01-01 23:54:26  read  country_2  2458153046  AdWords         Africa
1786  2018-01-01 23:54:39  read  country_6  2458153047   Reddit           Asia
1787  2018-01-01 23:54:45  read  country_2  2458153048   Reddit           Asia
1788  2018-01-01 23:55:04  read  country_7  2458153049  AdWords         Europe
1789  2018-01-01 23:56:42  read  country_4  2458153050  AdWords           Asia
1790  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1792  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1793  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1794  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia


The following code snippet on the other hand displays 10 random observations from our data set.

In [19]:
print(data.sample(10))

              my_datetime event    country     user_id   source          topic
680   2018-01-01 09:20:32  read  country_7  2458151941  AdWords         Europe
416   2018-01-01 05:44:15  read  country_7  2458151677   Reddit  North America
1713  2018-01-01 23:04:51  read  country_5  2458152974   Reddit  South America
76    2018-01-01 01:10:48  read  country_7  2458151337   Reddit           Asia
12    2018-01-01 00:09:11  read  country_5  2458151273   Reddit           Asia
541   2018-01-01 07:25:55  read  country_2  2458151802  AdWords  South America
976   2018-01-01 13:04:28  read  country_6  2458152237      SEO         Europe
1067  2018-01-01 14:27:12  read  country_7  2458152328   Reddit  South America
215   2018-01-01 03:02:32  read  country_2  2458151476   Reddit           Asia
1192  2018-01-01 16:09:33  read  country_5  2458152453      SEO  North America


### Selecting Specific columns from the data frame
The following code snippet selects specific features from our data set using square bracket notation.

In [20]:
# Selecting the country and topic columns only.
sub_data = data[["country", "topic"]]
print(sub_data)

        country          topic
0     country_7  North America
1     country_7  South America
2     country_7         Africa
3     country_7         Europe
4     country_8  North America
...         ...            ...
1790  country_2  North America
1791  country_8           Asia
1792  country_6           Asia
1793  country_7         Europe
1794  country_5           Asia

[1795 rows x 2 columns]


The following code snippet selects a single feature from our data set using dot notation.

In [21]:
sub_data = data.country
print(sub_data)

0       country_7
1       country_7
2       country_7
3       country_7
4       country_8
          ...    
1790    country_2
1791    country_8
1792    country_6
1793    country_7
1794    country_5
Name: country, Length: 1795, dtype: object


### Filtering specific observations from the data frame
The following code snippet filters the data frame, selecting only those data frames where the source feature has a value of SEO.

In [22]:
sub_data = data[data.source=="SEO"]
print(sub_data)

              my_datetime event    country     user_id source          topic
0     2018-01-01 00:01:01  read  country_7  2458151261    SEO  North America
1     2018-01-01 00:03:20  read  country_7  2458151262    SEO  South America
11    2018-01-01 00:08:57  read  country_7  2458151272    SEO      Australia
15    2018-01-01 00:11:22  read  country_7  2458151276    SEO  North America
16    2018-01-01 00:13:05  read  country_8  2458151277    SEO  North America
...                   ...   ...        ...         ...    ...            ...
1772  2018-01-01 23:45:58  read  country_7  2458153033    SEO  South America
1777  2018-01-01 23:49:52  read  country_5  2458153038    SEO  North America
1779  2018-01-01 23:51:25  read  country_4  2458153040    SEO  South America
1784  2018-01-01 23:54:03  read  country_2  2458153045    SEO  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052    SEO           Asia

[346 rows x 6 columns]


### Combining Pandas methods and operators
The following code snippet selects the first five observations using the head() method, and selects only the country and source features using the square bracket syntax.

In [23]:
sub_data = data.head(5)[["country", "source"]]
print(sub_data)

     country   source
0  country_7      SEO
1  country_7      SEO
2  country_7  AdWords
3  country_7  AdWords
4  country_8   Reddit


The following code snippet selects the country and source features from the data set first and then selects the first five observations using the head() method.

In [24]:
sub_data = data[["country", "source"]].head(5)
print(sub_data)

     country   source
0  country_7      SEO
1  country_7      SEO
2  country_7  AdWords
3  country_7  AdWords
4  country_8   Reddit


## Pandas Tutorial Challenge
Select the user_id, the country and the topic columns for the users who are from country_2! Print the first five rows only!

In [25]:
sub_data = data[data.country=="country_2"][["user_id", "country", "topic"]].head()
print(sub_data)

       user_id    country   topic
6   2458151267  country_2  Europe
13  2458151274  country_2  Europe
17  2458151278  country_2    Asia
19  2458151280  country_2    Asia
20  2458151281  country_2    Asia


# Website 3 Solutions

### Creating an empty pandas data frame
The following code snippet creates an empty data frame.

In [26]:
data = pd.DataFrame()
print(data)

Empty DataFrame
Columns: []
Index: []


### Creating a data frame from a list
The following code snippet creates a one-dimensional data frame from a list.

In [27]:
data = ["Google", "Bloomberg", "Meta", "Apple", "Tesla"] # the list
df = pd.DataFrame(data, columns=["name"])
print(df)

        name
0     Google
1  Bloomberg
2       Meta
3      Apple
4      Tesla


The following code snippet creates a multidimensional data frame from a list of lists.

In [28]:
data = [["Google", 1.27e12], ["Bloomberg", 123e9], ["Meta", 342.9e9], ["Apple", 2.23e12], ["Tesla", 649.99e9]] # the list
df = pd.DataFrame(data, columns=["name", "market cap"])
print(df)

        name    market cap
0     Google  1.270000e+12
1  Bloomberg  1.230000e+11
2       Meta  3.429000e+11
3      Apple  2.230000e+12
4      Tesla  6.499900e+11


The following code snippet creates a multidimensional data frame from a list of dictionaries

In [29]:
data = [{"name": "Google","market cap": 1.27e12}, {"name":"Bloomberg", "market cap": 1.23e9}, {"name":"Meta", "market cap":3.429e11}, {"name": "Apple", "market cap": 2.23e12}, {"name": "Tesla", "market cap": 6.4999e11}]
df = pd.DataFrame(data)
print(df)

        name    market cap
0     Google  1.270000e+12
1  Bloomberg  1.230000e+09
2       Meta  3.429000e+11
3      Apple  2.230000e+12
4      Tesla  6.499900e+11


### Creating a data frame from a dictionary
The following code snippet creates a data frame from a dictionary.
NB: The keys of the dictionary become feature labels and the values become observations for the corresponding feature.

In [30]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],'Age':[28,34,29,42]}
df = pd.DataFrame(data)
print(df)

    Name  Age
0    Tom   28
1   Jack   34
2  Steve   29
3  Ricky   42


In the above snippet, the rows of the data frame are numbered from 0 to n-1 by default, where n is the numbber of observations in the data frame.
In the code snippet below, the index argument of the data frame constructor will be used to label the rows.

In [31]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],'Age':[28,34,29,42]}
df = pd.DataFrame(data, index=["A1", "A2", "A3", "A4"])
print(df)

     Name  Age
A1    Tom   28
A2   Jack   34
A3  Steve   29
A4  Ricky   42


The following code snippet creates a dataframe from a dictionary of pandas series.

In [32]:
names = pd.Series(["Google", "Bloomberg", "Meta", "Apple", "Tesla"])
market_caps = pd.Series([1.27e12, 1.23e9, 3.429e11, 2.23e12, 6.4999e11])
data = {"name": names, "market_cap": market_caps}
df = pd.DataFrame(data)
print(df)

        name    market_cap
0     Google  1.270000e+12
1  Bloomberg  1.230000e+09
2       Meta  3.429000e+11
3      Apple  2.230000e+12
4      Tesla  6.499900e+11


### Adding columns to a data frame
The following code snippet adds a column to the df dataframe from a newly created data series.

In [33]:
new_series = pd.Series([1998, 1981, 2004, 1976, 2003])
df["year_founded"] = new_series
print(df)

        name    market_cap  year_founded
0     Google  1.270000e+12          1998
1  Bloomberg  1.230000e+09          1981
2       Meta  3.429000e+11          2004
3      Apple  2.230000e+12          1976
4      Tesla  6.499900e+11          2003


The following code snippet adds a column to the df column from an existing data frame.

In [34]:
df2 = df
df["random_factor"] = df["market_cap"] / df["year_founded"]
print(df)

        name    market_cap  year_founded  random_factor
0     Google  1.270000e+12          1998   6.356356e+08
1  Bloomberg  1.230000e+09          1981   6.208985e+05
2       Meta  3.429000e+11          2004   1.711078e+08
3      Apple  2.230000e+12          1976   1.128543e+09
4      Tesla  6.499900e+11          2003   3.245082e+08


### Deleting columns from a data frame
The following code snippet deletes a column from the data frame using the del python keyword.

In [35]:
print("-------before-deletion--------")
print(df)

# deletion
copy = df["random_factor"] # creating a copy of column to be deleted.

del df["random_factor"]

print("--------after-deletion---------")
print(df)
df["random_factor"] = copy

-------before-deletion--------
        name    market_cap  year_founded  random_factor
0     Google  1.270000e+12          1998   6.356356e+08
1  Bloomberg  1.230000e+09          1981   6.208985e+05
2       Meta  3.429000e+11          2004   1.711078e+08
3      Apple  2.230000e+12          1976   1.128543e+09
4      Tesla  6.499900e+11          2003   3.245082e+08
--------after-deletion---------
        name    market_cap  year_founded
0     Google  1.270000e+12          1998
1  Bloomberg  1.230000e+09          1981
2       Meta  3.429000e+11          2004
3      Apple  2.230000e+12          1976
4      Tesla  6.499900e+11          2003


The following code snippet deletes a column from the data frame using the data frame pop() method.

In [36]:
print("-------before-deletion--------")
print(df)

df.pop("random_factor")

print("--------after-deletion---------")
print(df)

-------before-deletion--------
        name    market_cap  year_founded  random_factor
0     Google  1.270000e+12          1998   6.356356e+08
1  Bloomberg  1.230000e+09          1981   6.208985e+05
2       Meta  3.429000e+11          2004   1.711078e+08
3      Apple  2.230000e+12          1976   1.128543e+09
4      Tesla  6.499900e+11          2003   3.245082e+08
--------after-deletion---------
        name    market_cap  year_founded
0     Google  1.270000e+12          1998
1  Bloomberg  1.230000e+09          1981
2       Meta  3.429000e+11          2004
3      Apple  2.230000e+12          1976
4      Tesla  6.499900e+11          2003


### Selecting rows and columns from a data frame with loc
The following code snippets selects the name and year_founded columns as well as the 0th, 2nd, and 4th row from the df data frame using loc.

In [37]:
selection = df.loc[[0,2,4], ["name", "year_founded"]]
print(selection)

     name  year_founded
0  Google          1998
2    Meta          2004
4   Tesla          2003


The following code snippet performs the above selection using iloc.

In [38]:
selection = df.iloc[[0, 2, 4], [0, 2]]
print(selection)

     name  year_founded
0  Google          1998
2    Meta          2004
4   Tesla          2003


The following code snippet selects a range of rows and columns from the df data frame using slicing.
NB: For iloc, stop index of the slice is exclusive, but for loc, both start and stop index of the slice are inclusive.

In [39]:
selection = df.iloc[0:2, 1:3]
print(selection)

     market_cap  year_founded
0  1.270000e+12          1998
1  1.230000e+09          1981


### Deleting rows from a data frame
Just as columns are deleted from a data frame using the del keyword and .pop method, Rows can be deleted using the del keyword and the .drop method.

The following code snippet deletes the 3rd column from the df data frame.

In [40]:
print("--------Before-deletion---------")
print(df)

df = df.drop(2)

print("---------After-deletion--------")
print(df)

--------Before-deletion---------
        name    market_cap  year_founded
0     Google  1.270000e+12          1998
1  Bloomberg  1.230000e+09          1981
2       Meta  3.429000e+11          2004
3      Apple  2.230000e+12          1976
4      Tesla  6.499900e+11          2003
---------After-deletion--------
        name    market_cap  year_founded
0     Google  1.270000e+12          1998
1  Bloomberg  1.230000e+09          1981
3      Apple  2.230000e+12          1976
4      Tesla  6.499900e+11          2003


# Website 4 Solutions

### Creating a data frame from scratch (using a dictionary)
The following code snippet creates a pandas data frame from scratch using the below dictionary.

In [41]:
data = {'apples': [3, 2, 0, 1], 'oranges': [0, 3, 7, 2]}
df = pd.DataFrame(data)
print(df)

   apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2


The following code snippet does the above, but includes row labels for the data frame.

In [42]:
data = {'apples': [3, 2, 0, 1], 'oranges': [0, 3, 7, 2]}
df = pd.DataFrame(data,index=["June", "Robert", "Lily", "David"])
print(df)

        apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2


In the above data frame, each row represents a particular customer's order.
### Selecting a single row
In the code snippet below, a single row(custmer) is selected via the iloc method and row label.

In [43]:
selection = df.loc["June"]
print(selection)

apples     3
oranges    0
Name: June, dtype: int64


### Reading data from a csv file
The following code snippet demonstrates how to use the read_csv method to read csv files into a pandas data frame.

In [44]:
df = pd.read_csv("c://users/franc/PycharmProjects/intro-to-data-science/data/purchases.csv", delimiter=",", index_col=0)
print(df)

        apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2


### Reading data from a json file
The following code snippet reads the purchases data from a json file instead via the read_json method.

In [45]:
df = pd.read_json("C:/users/franc/PycharmProjects/intro-to-data-science/data/purchases.json")
print(df)


           apples  oranges
0   David       3        0
1    June       2        3
2    Lily       0        7
3  Robert       1        2


### Reading data from an sql database.
Pandas can read tables stored inside a relational database and store them as data frames.
The following code snippets demonstrates how to read data into a pandas data frame from an sql database. <br>
For the purposes of this tutorial, we'll be making use of the sqlite database which stores databases as a single file with the extension .db from which we can read data similar to how we would read data from a csv file.

We first need to install <strong>pysqlite3</strong> which is a database driver that allows python programs to interact with sqlite databases using the following command:

In [46]:
!pip install pysqlite3

  Using cached pysqlite3-0.4.7.tar.gz (40 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py clean for pysqlite3
Failed to build pysqlite3
  Running setup.py install for pysqlite3: started
  Running setup.py install for pysqlite3: finished with status 'done'


  Command arguments: 'C:\Users\franc\anaconda3\envs\intro-to-data-science\python.exe' -u -c '
  exec(compile('"'"''"'"''"'"'
  # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
  #
  # - It imports setuptools before invoking setup.py, to enable projects that directly
  #   import from `distutils.core` to work with newer packaging standards.
  # - It provides a clear error message when setuptools is not installed.
  # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
  #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
  #     manifest_maker: standard file '"'"'-c'"'"' not found".
  # - It generates a shim setup.py, for handling setup.cfg-only projects.
  import os, sys, tokenize
  
  try:
      import setuptools
  except ImportError as error:
      print(
          "ERROR: Can not execute `setup.py` since setuptools is not available in "
          "the build environment.",
          file=sys.

Then we have to create a connection to our sqlite database using the connect method of the sqlite3 driver.

In [47]:
import sqlite3

connection = sqlite3.connect("C:/users/franc/PycharmProjects/intro-to-data-science/data/data.db")
# connection is now a database connection object we can use to access data stored in the data.db database.

The following command uses the read_sql_query() method to query the data.db database through our connection object, and read the resulting table as a pandas data frame which we store in the variable df.

In [48]:
df = pd.read_sql_query("SELECT * FROM purchases", connection)
print(df)

    index  apples  oranges
0    June       3        0
1  Robert       2        3
2    Lily       0        7
3   David       1        2


We failed to state that the index column of the purchases table in our data.db database was to be treated as the index column. <br>
But all hope is not lost as we can use the set_index() data frame method to set our index column.

In [49]:
df = df.set_index("index")
print(df)

        apples  oranges
index                  
June         3        0
Robert       2        3
Lily         0        7
David        1        2


### Converting a pandas data frame into other formats.
The following code snippet converts the df data frame into a **csv file** stored in the <em>output</em> sub-directory of our <em>data</em> directory.

In [50]:
df.to_csv("../data/output/output_csv.csv")

The following code snippet converts the df data frame into a **json file** and stores it in the <em>output</em> sub-directory of the <em>data</em> directory.

In [51]:
df.to_json("../data/output/output_json.json")

The following code snippet converts the df data frame into an **sql table** and stores it in the data.db sqlite database via our connection object.

In [52]:
df.to_sql("new purchases", connection)

ValueError: Table 'new purchases' already exists.

## The following section covers some important data frame operations
To demonstrate various operations that can be performed on a pandas database, we will be using a [movies data set](https://storage.googleapis.com/kagglesdsdata/datasets/516826/954748/df_movie_ratings.csv?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20221015%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20221015T113249Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=2736f429dbdfa2c10167ce36777a57e8f2874890ca7396cae864f32bdde4b06a2e1f872cec927266e26740a615c87eea111bb4e6f771c2b2d440bcdf6e22548ddafb32754312dd011087dafc8eb5bca0ecff86b3752e3b834498ba96327548216bf36d9d3a29e4243d95d3052c5abe1436b41027d40ab7743ae73f539b620b32473ddd71dd4e3f3e9a93e86b1fc24a02286786f8004719a68e0fb720b0d76b0c1194d21b241d91e96ae5eb345c9d4393b3a98a5742754d6a6739349a994d4cfb59a51dc308e10c9f6ebd2a9123968af54150de38e149c2318b9a9226607c003f2a7bcdc9f30cc1b4e98c0083f5c84f54e8663bbdcf1e80d672d60ff737a49c53) we downloaded from kaggle.

In [53]:
# importing the movies data set from a csv file
movies= pd.read_csv("../data/df_movie_ratings.csv", delimiter=",")
print(movies)

      Unnamed: 0                                       movie  year  imdb  \
0              0                                   Gladiator  2000   8.5   
1              1                                     Memento  2000   8.4   
2              2                                      Snatch  2000   8.3   
3              3                         Requiem for a Dream  2000   8.3   
4              4                                       X-Men  2000   7.4   
...          ...                                         ...   ...   ...   
3611        3611                             Triple Frontier  2019   6.4   
3612        3612  How to Train Your Dragon: The Hidden World  2019   7.5   
3613        3613                 Men in Black: International  2019   5.6   
3614        3614                      Zombieland: Double Tap  2019   6.8   
3615        3615                              Murder Mystery  2019   6.0   

      metascore    votes                                   genre  runtime  \
0         

### Viewing the data
The following snippets cover various data frame methods that may be useful for viewing or analyzing the data stored in the data frame.

**Using the head() method to view first few observations in data frame:**

In [54]:
data = movies.head(10) # Selects first 10 observations in data frame.
print(data)

   Unnamed: 0                   movie  year  imdb  metascore    votes  \
0           0               Gladiator  2000   8.5         67  1265139   
1           1                 Memento  2000   8.4         80  1065249   
2           2                  Snatch  2000   8.3         55   744029   
3           3     Requiem for a Dream  2000   8.3         68   725875   
4           4                   X-Men  2000   7.4         64   550968   
5           5               Cast Away  2000   7.8         73   491129   
6           6         American Psycho  2000   7.6         64   453294   
7           7             Unbreakable  2000   7.3         62   370098   
8           8  Mission: Impossible II  2000   6.1         59   299974   
9           9        Meet the Parents  2000   7.0         73   299610   

                                   genre  runtime        gross  n_imdb  
0     ['Action', ' Adventure', ' Drama']  155 min  187705427.0    85.0  
1               ['Mystery', ' Thriller']  113 min 

**Using the tail() method to view last few observations in data frame:**

In [55]:
data = movies.tail(3) # Selects last 3 observations in data frame.
print(data)

      Unnamed: 0                        movie  year  imdb  metascore  votes  \
3613        3613  Men in Black: International  2019   5.6         38  85672   
3614        3614       Zombieland: Double Tap  2019   6.8         55  82913   
3615        3615               Murder Mystery  2019   6.0         38  81450   

                                    genre  runtime       gross  n_imdb  
3613  ['Action', ' Adventure', ' Comedy']  114 min  79800736.0    56.0  
3614     ['Action', ' Comedy', ' Horror']   99 min  26803104.0    68.0  
3615      ['Action', ' Comedy', ' Crime']   97 min         NaN    60.0  


**Using the info() method to obtain general information about data frame:**
The info() method displays number of rows, columns, non-null values in each column, data type of values in each column, as well as overall memory usage of the data frame.

In [56]:
print(movies.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3616 entries, 0 to 3615
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  3616 non-null   int64  
 1   movie       3616 non-null   object 
 2   year        3616 non-null   int64  
 3   imdb        3616 non-null   float64
 4   metascore   3616 non-null   int64  
 5   votes       3616 non-null   int64  
 6   genre       3616 non-null   object 
 7   runtime     3616 non-null   object 
 8   gross       3536 non-null   float64
 9   n_imdb      3616 non-null   float64
dtypes: float64(3), int64(4), object(3)
memory usage: 282.6+ KB
None


**Using the describe() method to retrieve quantitative information about dataframe:**
The describe() method can be used to retrieve the number of observations, average, standard deviation, maximum, minimum, 25th, 50th, and 75th percentile of the data in each column of a dataframe as follows:

In [57]:
print(movies.describe())

        Unnamed: 0         year         imdb    metascore         votes  \
count  3616.000000  3616.000000  3616.000000  3616.000000  3.616000e+03   
mean   1807.500000  2009.355088     7.114934    63.591814  3.275271e+05   
std    1043.993614     5.756038     0.738934    15.675350  2.261506e+05   
min       0.000000  2000.000000     4.100000    24.000000  8.145000e+04   
25%     903.750000  2004.000000     6.600000    52.000000  1.874390e+05   
50%    1807.500000  2009.000000     7.200000    64.000000  2.527240e+05   
75%    2711.250000  2014.000000     7.700000    75.000000  3.997345e+05   
max    3615.000000  2019.000000     9.000000   100.000000  2.173427e+06   

              gross       n_imdb  
count  3.536000e+03  3616.000000  
mean   1.269317e+08    71.149336  
std    1.175377e+08     7.389338  
min    3.600000e+03    41.000000  
25%    4.724575e+07    66.000000  
50%    9.354176e+07    72.000000  
75%    1.710725e+08    77.000000  
max    9.366622e+08    90.000000  


Note that all categorical variables had their columns ignored.
But `describe()` can still give information about categorical variables as shown below:

In [58]:
print(movies.movie.describe()) # Calls describe() method on categorical variable of the dataframe, movie.

count          3616
unique          904
top       Gladiator
freq              4
Name: movie, dtype: object


From the output, we can tell that the movie column has only 904 unique values, the most frequent movie title is "Gladiator", and it occurred 4 times in the movie column.

**Using the shape attribute to determine number of rows and columns in data frame:**

In [59]:
shape = movies.shape
print(shape)
print("No. of rows:", shape[0])
print("No. of columns:", shape[1])

(3616, 10)
No. of rows: 3616
No. of columns: 10


**Using the value_counts() method to count the number of times each value of a particular column occurs:**

In [60]:
movie = movies.movie # selecting the movie column of the movies dataframe.
print(movie.value_counts())

Gladiator                                4
We're the Millers                        4
Lo imposible                             4
Project X                                4
The Wolf of Wall Street                  4
                                        ..
The Holiday                              4
The Fast and the Furious: Tokyo Drift    4
Crank                                    4
Perfume: The Story of a Murderer         4
Murder Mystery                           4
Name: movie, Length: 904, dtype: int64


**Using the corr() method to find the relationship between numerical variables in a dataframe**
The dataframe `corr()` method returns the correlation between all quantitative features in the dataframe as a dataframe as shown below:

In [61]:
correlation = movies.corr()
print(correlation)

            Unnamed: 0      year      imdb  metascore     votes     gross  \
Unnamed: 0    1.000000  0.998707  0.008577   0.092386 -0.036715  0.190007   
year          0.998707  1.000000  0.014492   0.096722 -0.028082  0.196261   
imdb          0.008577  0.014492  1.000000   0.699024  0.513588  0.032634   
metascore     0.092386  0.096722  0.699024   1.000000  0.324337  0.070391   
votes        -0.036715 -0.028082  0.513588   0.324337  1.000000  0.443705   
gross         0.190007  0.196261  0.032634   0.070391  0.443705  1.000000   
n_imdb        0.008577  0.014492  1.000000   0.699024  0.513588  0.032634   

              n_imdb  
Unnamed: 0  0.008577  
year        0.014492  
imdb        1.000000  
metascore   0.699024  
votes       0.513588  
gross       0.032634  
n_imdb      1.000000  


As is expected, each column has a correlation of 1 with itself.
Another thing worth taking note of is that `imdb` seems to have a strong positive correlation with `metascore` (ie. 0.699024). This means that the higher the metascore of a movie, the higher its imdb rating.

### Removing duplicate observations from data frame
The **drop_duplicates()** method is used to remove duplicate rows from our dataframe.
Since no duplicates exist in our data frame, we will create a new data frame consisting of our movies data frame concatenated with itself, which will therefore contain a duplicate entry for each entry in the movies data frame.

In [62]:
copy = pd.concat([movies, movies])

We will use the shape attribute to confirm that the copy data frame contains twice as many observations as the movies data frame.

In [63]:
print("No. of rows in movies:",movies.shape[0])
print("No. of rows in copy:",copy.shape[0])

No. of rows in movies: 3616
No. of rows in copy: 7232


Now we will use the drop_duplicates() method to remove duplicate entries from our copy data frame.

In [64]:
copy.drop_duplicates(inplace=True, keep="first")

The **inplace** argument specifies that we want to remove the duplicates from the data frame on which the method was called rather than return a copy of that data frame with the duplicates removed.
The **keep** argument also specifies in this case that we want to keep the first occurrence of the duplicate and remove all other occurrences.

To confirm that all duplicates have been removed from our copy data frame, we will again use the shape method and compare number of rows of copy with number of rows of movies.

In [65]:
print("No. of rows in movies:",movies.shape[0])
print("No. of rows in copy:",copy.shape[0])
print("Voila🎉🎉")

No. of rows in movies: 3616
No. of rows in copy: 3616
Voila🎉🎉


### Modifying column labels of the data frame.
The **columns** attribute returns an iterable of all the column labels in a data frame and can likewise be used to modify column labels.

**Displaying all column labels of the movies data frame:**

In [66]:
columns = movies.columns
print(columns)

Index(['Unnamed: 0', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre',
       'runtime', 'gross', 'n_imdb'],
      dtype='object')


**Changing all column labels of data frame via lists**

In [67]:
copy = movies.copy() # Creating a copy of movies so we can use the original column labels later
new_columns = ["", "new_movie", "new_year", "new_imdb", "new_metascore", "new_votes", "new_genre", "new_runtime", "new_gross", "new_n_imdb"]
print("----------before-modification----------")
print(copy.columns)

copy.columns = new_columns
print("----------after-modification-----------")
print(copy.columns)

----------before-modification----------
Index(['Unnamed: 0', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre',
       'runtime', 'gross', 'n_imdb'],
      dtype='object')
----------after-modification-----------
Index(['', 'new_movie', 'new_year', 'new_imdb', 'new_metascore', 'new_votes',
       'new_genre', 'new_runtime', 'new_gross', 'new_n_imdb'],
      dtype='object')


**Changing all column labels of dataframe via list comprehensions**

In [68]:
print("-----before-alteration------")
print(copy.columns)

copy.columns = [column[4:] for column in copy.columns]
print("-----after-alternation------")
print(copy.columns)

-----before-alteration------
Index(['', 'new_movie', 'new_year', 'new_imdb', 'new_metascore', 'new_votes',
       'new_genre', 'new_runtime', 'new_gross', 'new_n_imdb'],
      dtype='object')
-----after-alternation------
Index(['', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre', 'runtime',
       'gross', 'n_imdb'],
      dtype='object')


**Changing all column labels of dataframe using rename method:**

In [69]:
print("-----before-modification-----")
print(copy.columns)

copy.rename(columns={"":"ID"}, inplace=True) # Changes empty string label of first column to "ID"

print("-----after-modification-----")
print(copy.columns)

-----before-modification-----
Index(['', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre', 'runtime',
       'gross', 'n_imdb'],
      dtype='object')
-----after-modification-----
Index(['ID', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre', 'runtime',
       'gross', 'n_imdb'],
      dtype='object')


### Identifying and fixing missing values in dataframe
Sometimes some data points may be missing for certain observations in our dataset.
Such data points are said to be null, and the following code snippets show how to deal with such cases in pandas.

**Detecting missing values in a data frame:**
The `isnull()` method returns a dataframe frame with all cells replaced with boolean values. Cells with a value of null are replaced with `True`, and cells with a value other than null are replaced with `False`.

In [70]:
null_frame = movies.isnull()
print(null_frame)

      Unnamed: 0  movie   year   imdb  metascore  votes  genre  runtime  \
0          False  False  False  False      False  False  False    False   
1          False  False  False  False      False  False  False    False   
2          False  False  False  False      False  False  False    False   
3          False  False  False  False      False  False  False    False   
4          False  False  False  False      False  False  False    False   
...          ...    ...    ...    ...        ...    ...    ...      ...   
3611       False  False  False  False      False  False  False    False   
3612       False  False  False  False      False  False  False    False   
3613       False  False  False  False      False  False  False    False   
3614       False  False  False  False      False  False  False    False   
3615       False  False  False  False      False  False  False    False   

      gross  n_imdb  
0     False   False  
1     False   False  
2     False   False  
3     False

The above result means that the cell on row `3615` and column `"gross"` is null.

**Identifying columns with missing values in a dataframe**
The `sum()` dataframe method sums up all values in each column whose values are summable and returns a data series of all the resulting sums.
Each value in the data series is indexed with the label of the column whose values they are the sum of.

The `sum()` method can be used in conjunction with the `isnull()` method to identify all columns with missing values as shown below.

In [71]:
null_columns = movies.isnull().sum()
print("The following are the columns in the movies dataframe with missing values and the number of missing values for each column:")
print(null_columns)

The following are the columns in the movies dataframe with missing values and the number of missing values for each column:
Unnamed: 0     0
movie          0
year           0
imdb           0
metascore      0
votes          0
genre          0
runtime        0
gross         80
n_imdb         0
dtype: int64


From the above results, it's safe to say that the gross column has 80 missing values whereas all other columns have no missing values.

**Removing all rows with missing or null values**
The **dropna()** method is used to remove all rows in a dataframe with missing values as shown below:

In [72]:
copy = movies.copy() # Creating a copy of the movies data frame.
print("No. of rows before deletion:", copy.shape[0])
copy.dropna(inplace=True)
print("No. of rows after deletion:", copy.shape[0])

No. of rows before deletion: 3616
No. of rows after deletion: 3536


From the above results, we can notice that the 80 rows that contained null values were deleted from the dataframe.

**Removing all columns with missing or null values**
The **dropna()** method can also be used to remove columns from a dataframe whose cells contain null values.
To accomplish this, the `axis` argument of the `dropna()` method is set to `1` as shown below:

In [73]:
copy = movies.copy() # Creating a copy of the movies data frame.
print("Columns before removal:\n", copy.columns)
copy.dropna(inplace=True, axis=1)
print("Columns after removal:\n", copy.columns)

Columns before removal:
 Index(['Unnamed: 0', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre',
       'runtime', 'gross', 'n_imdb'],
      dtype='object')
Columns after removal:
 Index(['Unnamed: 0', 'movie', 'year', 'imdb', 'metascore', 'votes', 'genre',
       'runtime', 'n_imdb'],
      dtype='object')


Notice that the `gross` column, which was the only column containing missing values, has been removed from our dataframe.

**Replacing all missing values in a dataframe**
Imputation refers to the replacement of missing values in a dataframe with other values as opposed to deleting them.
The **fillna()** method is used for imputation in pandas.

The `fillna()` method replaces all cells in the dataframe that have missing values with a specified value.
In the code snippet below we replace all the missing values in a copy of the `movies` dataframe with the mean of the values in the gross column:

In [74]:
copy = movies.copy() # Creating a copy of the movies dataframe.
average = copy.gross.mean()
print("------before-imputation------")
print(copy.isnull().sum())

# Replacing null values with average
copy.fillna(average, inplace=True)
print("------after-imputation-------")
print(copy.isnull().sum())

------before-imputation------
Unnamed: 0     0
movie          0
year           0
imdb           0
metascore      0
votes          0
genre          0
runtime        0
gross         80
n_imdb         0
dtype: int64
------after-imputation-------
Unnamed: 0    0
movie         0
year          0
imdb          0
metascore     0
votes         0
genre         0
runtime       0
gross         0
n_imdb        0
dtype: int64


Note that after imputation, all null values in the `gross` column were replaced with the `average`, and the `gross` column therefore no longer has missing values.

## DataFrame slicing, selecting, extracting
In extracting data from a dataframe, a data scientist's goal is usually one of the following:
   - Extracting a single cell
   - Extracting a single column
   - Extracting a single row
   - Extracting multiple columns
   - Extracting multiple rows

With the above actions, one can extract any piece of data they require from a dataframe, be it a single value, an entire row, an entire column, a small grid wtihin the dataframe, etc.

### Requirements for extracting data from a dataframe
Python provides three useful tools for extracting data from your dataframe:
1. The index operator `[]`
2. The loc dataframe attribute `.loc`
3. The iloc dataframe attribute `.iloc`

In addition to this, every dataframe has **row labels** and **column labels**, which must be used in conjunction with the above 3 tools in order to extract data from a dataframe.

There are 2 kinds of row and column labels in the dataframe:
 1. The actual label
 2. The integer-location-based label

The **actual labels** refer to the real labels of the rows and columns that come with the dataframe.
 - These labels may be strings, integers, or any other data type.
 - In the `movies` dataframe for instance, the actual column labels include `"movie"`, `"genre"`, `"rating"`, etc., which are all strings.
  - The actual row labels are the integers from `0` to `3616`, but they can actually be of any datatype.
    - Dataframes must always have row and column labels, and when a dataframe you create or read does not have row labels for instance, pandas automatically assigns integer labels to the dataframe which is the case with our movies dataframe.

The **integer-location-based labels** are unreal labels pandas assigns to each row and column based on its position in the dataframe.
 - The first row or column is given an integer-location-based label of `0` due to its position as the first row/column in the dataframe.
 - The next column or row is given an integer-location based label of `1`, and so on and so forth.
 - These unreal labels will always be integers, and are different from the actual labels on the dataframe which may be any data type.

The movies dataframe is printed below for reference.

In [75]:
print(movies)

      Unnamed: 0                                       movie  year  imdb  \
0              0                                   Gladiator  2000   8.5   
1              1                                     Memento  2000   8.4   
2              2                                      Snatch  2000   8.3   
3              3                         Requiem for a Dream  2000   8.3   
4              4                                       X-Men  2000   7.4   
...          ...                                         ...   ...   ...   
3611        3611                             Triple Frontier  2019   6.4   
3612        3612  How to Train Your Dragon: The Hidden World  2019   7.5   
3613        3613                 Men in Black: International  2019   5.6   
3614        3614                      Zombieland: Double Tap  2019   6.8   
3615        3615                              Murder Mystery  2019   6.0   

      metascore    votes                                   genre  runtime  \
0         

Note that the actual row labels for the `movies` dataframe are integers.

To make it easier to understand the difference between actual labels of a dataframe and integer-location based labels, we will use the values in the `movie` column as the new row labels via the `set_index()` method as shown below.

In [80]:
copy = movies.copy() # Copying movies dataframe

copy.set_index("movie", inplace=True) # Using the movie column as row labels.
print(copy)

                                            Unnamed: 0  year  imdb  metascore  \
movie                                                                           
Gladiator                                            0  2000   8.5         67   
Memento                                              1  2000   8.4         80   
Snatch                                               2  2000   8.3         55   
Requiem for a Dream                                  3  2000   8.3         68   
X-Men                                                4  2000   7.4         64   
...                                                ...   ...   ...        ...   
Triple Frontier                                   3611  2019   6.4         61   
How to Train Your Dragon: The Hidden World        3612  2019   7.5         71   
Men in Black: International                       3613  2019   5.6         38   
Zombieland: Double Tap                            3614  2019   6.8         55   
Murder Mystery              

### Extracting data from a dataframe without loc and iloc
The following snippets demonstrate how we can select single cells, single rows, single columns, and multiple rows and columns from our dataframe without using the loc and iloc attributes.

**Using the index operator to extract a single cell from your dataframe**
A cell in a dataframe is an intersection of a row and a column.
To extract a single cell in the dataframe, you need the row label and column label for the row and column on which the cell lies respectively.
> dataframe[column_label][row_label]

The code cell below selects the cell located at the intersection of the `genre` column and `Gladiator`

In [76]:
selection = copy["genre"]["Gladiator"]
print("")

### Indexing a dataframe
Slice indexing re
There are two types of indexing:
 - Slice indexing
 - Positional Indexing


In [77]:
print(copy.head().loc["Gladiator", "genre"])
print(copy.head()["genre"]["Gladiator"])

KeyError: 'Gladiator'

In [ ]:
# Without loc and iloc pandas attributes.
# Introduce the loc and iloc pandas attributes. The loc and iloc pandas attributes allow you to select rows and columns by label. The loc uses the specific row and column labels where as the iloc uses integer arguments it assigns to each row and to each column.